internal/controllers/gateway/gateway_controller.go

```diff
@@ -89,67 +89,49 @@ func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Resource: "referencegrants",
 	})
 
-	// generate the controller object and attach it to the manager and link the reconciler object
-	c, err := controller.New("gateway-controller", mgr, controller.Options{
-		Reconciler: r,
-		LogConstructor: func(_ *reconcile.Request) logr.Logger {
-			return r.Log
-		},
-		CacheSyncTimeout: r.CacheSyncTimeout,
-	})
-	if err != nil {
-		return err
-	}
-
-	// watch Gateway objects, filtering out any Gateways which are not configured with
-	// a supported GatewayClass controller name.
-	if err := c.Watch(
-		source.Kind(mgr.GetCache(), &gatewayapi.Gateway{}),
-		&handler.EnqueueRequestForObject{},
-		predicate.NewPredicateFuncs(r.gatewayHasMatchingGatewayClass),
-	); err != nil {
-		return err
-	}
-
-	// watch for updates to gatewayclasses, if any gateway classes change, enqueue
-	// reconciliation for all supported gateway objects which reference it.
-	if err := c.Watch(
-		source.Kind(mgr.GetCache(), &gatewayapi.GatewayClass{}),
-		handler.EnqueueRequestsFromMapFunc(r.listGatewaysForGatewayClass),
-		predicate.NewPredicateFuncs(r.gatewayClassMatchesController),
-	); err != nil {
-		return err
-	}
-
-	// if an update to the gateway service occurs, we need to make sure to trigger
-	// reconciliation on all Gateway objects referenced by it (in the most common
-	// deployments this will be a single Gateway).
-	if err := c.Watch(
-		source.Kind(mgr.GetCache(), &corev1.Service{}),
-		handler.EnqueueRequestsFromMapFunc(r.listGatewaysForService),
-		predicate.NewPredicateFuncs(r.isGatewayService),
-	); err != nil {
-		return err
-	}
-
-	// if a HTTPRoute gets accepted by a Gateway, we need to make sure to trigger
-	// reconciliation on the gateway, as we need to update the number of attachedRoutes.
-	if err := c.Watch(
-		source.Kind(mgr.GetCache(), &gatewayapi.HTTPRoute{}),
-		handler.EnqueueRequestsFromMapFunc(r.listGatewaysForHTTPRoute),
-	); err != nil {
-		return err
-	}
+	blder := ctrl.NewControllerManagedBy(mgr).
+		// set the controller name
+		Named("gateway-controller").
+		// set the controller options
+		WithOptions(controller.Options{
+			LogConstructor: func(_ *reconcile.Request) logr.Logger {
+				return r.Log
+			},
+			CacheSyncTimeout: r.CacheSyncTimeout,
+		}).
+		// watch Gateway objects, filtering out any Gateways which are not configured with
+		// a supported GatewayClass controller name.
+		For(&gatewayapi.Gateway{}).
+		WithEventFilter(predicate.NewPredicateFuncs(r.gatewayHasMatchingGatewayClass)).
```

Comment:
Changing this fixes the `envtest` suite. Haven't checked the integration suite but probably that will fix it as well.

```suggestion
		For(&gatewayapi.Gateway{},
				builder.WithPredicates(predicate.NewPredicateFuncs(r.gatewayHasMatchingGatewayClass)),
		).
```

Reply:
Thank you!!